pkg/k8shandler/cluster.go

```diff
@@ -315,6 +368,68 @@ func (cState *ClusterState) amendPods(dpl *v1alpha1.Elasticsearch) error {
 	return nil
 }
 
+func upgradeInProgress(dpl *v1alpha1.Elasticsearch) *v1alpha1.ElasticsearchNodeStatus {
+	for _, node := range dpl.Status.Nodes {
+		if node.UnderUpgrade {
+			return &node
+		}
+	}
+	return nil
+}
+
+func selectNodeForUpgrade(dpl *v1alpha1.Elasticsearch) *v1alpha1.ElasticsearchNodeStatus {
+	for _, node := range dpl.Status.Nodes {
+		if !node.UnderUpgrade {
+			return &node
+		}
+	}
+	return nil
+}
+
+func (cState *ClusterState) beginUpgrade(dpl *v1alpha1.Elasticsearch, owner metav1.OwnerReference) (*v1alpha1.ElasticsearchNodeStatus, error) {
+	masterPod, err := getRunningMasterPod(dpl.Name, dpl.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	if err = utils.DisableShardAllocation(masterPod); err != nil {
+		return nil, err
+	}
+	dpl.Status.ShardAllocationEnabled = false
+	if err = utils.PerformSyncedFlush(masterPod); err != nil {
+		return nil, err
+	}
+	node, err := cState.upgradeNode(dpl, owner)
+	if err != nil {
+		return nil, err
+	}
+	err = sdk.Update(dpl)
```

Comment:
you'll have to use `retry.RetryOnConflict` here - see https://github.com/openshift/elasticsearch-operator/pull/42

In fact - you may have to do an update after each modification of the `dpl.Status`:

```go
	if err = utils.DisableShardAllocation(masterPod); err != nil {
		return nil, err
	} 
       retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
	 if getErr := sdk.Get(dpl); getErr != nil {
           return getErr
         }
         dpl.Status.ShardAllocationEnabled = false // maybe check to see if it is false already?
         return sdk.Update(dpl)
     })
```

Reply:
hmm - It's going to be tricky - what if the `utils.DisableShardAllocation` succeeds but the `cState.upgradeNode` fails?  I notice that you do not handle the error return from `beginUpgrade` in `restartCluster`. I wonder if we need some sort of transactional semantics with rollback in case one of the steps fails?  For example, if `upgradeNode` fails, how do you know you need to re-enable shard allocation?  Is this a case where we may need to use `defer` to "stack" the unwind functions?